extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js

```diff
@@ -105,45 +109,213 @@ export class AmpStoryShoppingAttachment extends AMP.BaseElement {
     }
     this.storeService_.subscribe(
       StateProperty.PAGE_ATTACHMENT_STATE,
-      (isOpen) => this.onPageAttachmentStateUpdate_(isOpen)
+      (isOpen) => {
+        const shoppingData = this.storeService_.get(
+          StateProperty.SHOPPING_DATA
+        );
+        this.checkClearActiveProductData_(isOpen, shoppingData);
+        this.updateTemplate_(isOpen, shoppingData);
+      }
+    );
+    this.storeService_.subscribe(
+      StateProperty.SHOPPING_DATA,
+      (shoppingData) => {
+        const isOpen = this.storeService_.get(
+          StateProperty.PAGE_ATTACHMENT_STATE
+        );
+        this.updateTemplate_(isOpen, shoppingData);
+      }
     );
   }
 
   /**
-   * On attachment state update, check if on active page and populate plp.
+   * Active product data is cleared if attachment is opening and activeProductData was previously set.
+   * It's cleared on open instead of close so that content doesn't jump while closing.
    * @param {boolean} isOpen
+   * @param {!Array<!ShoppingConfigDataDef>} shoppingData
    * @private
    */
-  onPageAttachmentStateUpdate_(isOpen) {
-    const isOnActivePage =
-      this.pageEl_.id === this.storeService_.get(StateProperty.CURRENT_PAGE_ID);
-
-    if (isOpen && isOnActivePage) {
-      this.populatePlp_();
+  checkClearActiveProductData_(isOpen, shoppingData) {
+    const {activeProductData} = shoppingData;
+    if (!isOpen && activeProductData && this.isOnActivePage_()) {
+      this.storeService_.dispatch(Action.ADD_SHOPPING_DATA, {
+        'activeProductData': null,
+      });
     }
   }
 
   /**
-   * Renders a list of the products on the page, knows as a "Product Listing Page" or PLP.
+   * On attachment state or shopping data state update, check if on active page and update template.
+   * @param {boolean} isOpen
+   * @param {!Array<!ShoppingConfigDataDef>} shoppingData
    * @private
    */
-  populatePlp_() {
-    if (this.plpContainer_.querySelector('.i-amphtml-amp-story-shopping-plp')) {
+  updateTemplate_(isOpen, shoppingData) {
+    if (!isOpen || !this.isOnActivePage_()) {
       return;
     }
-    const shoppingData = this.storeService_.get(StateProperty.SHOPPING_DATA);
     const shoppingDataForPage = this.shoppingTags_.map(
       (shoppingTag) => shoppingData[shoppingTag.getAttribute('data-product-id')]
     );
 
-    const plp = this.renderPlpTemplate_(shoppingDataForPage);
-    this.mutateElement(() => {
-      this.plpContainer_.appendChild(plp);
+    // If there is only one product on the page, feature it.
+    // Otherwise feature the active product.
+    const singleProductOnPage =
+      shoppingDataForPage.length === 1 && shoppingDataForPage[0];
+    const featuredProduct =
+      shoppingData.activeProductData || singleProductOnPage;
```

Comment:
Is there any scenario where `activeProductData` would not be `singleProductOnPage`? Is it when a user tries to open a PLP with 1 product that you force it into a PDP? (if yes, where does this happen in the code?)

Reply:
Yes. That happens in `buildTemplate`. I renamed the variables to try and make this more clear.